File: src/main/java/nl/dannyjelsma/soundboardbot/audio/SoundManager.java
package nl.dannyjelsma.soundboardbot.audio;

import com.sedmelluq.discord.lavaplayer.player.AudioPlayer;
import com.sedmelluq.discord.lavaplayer.player.AudioPlayerManager;
import com.sedmelluq.discord.lavaplayer.player.DefaultAudioPlayerManager;
import com.sedmelluq.discord.lavaplayer.source.AudioSourceManagers;
import com.sedmelluq.discord.lavaplayer.track.AudioTrack;

import java.io.File;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class SoundManager {

    private final AudioPlayer audioPlayer;
    private static SoundManager instance;
    private final List<Sound> sounds;
    private final AudioPlayerManager audioPlayerManager;
    private final SoundScheduler scheduler;
    private boolean shuffle = false;
    private boolean ttsAnnouncement = true;
    private final HashMap<String, AudioTrack> songCache = new HashMap<>();
    private boolean loop = false;

    private SoundManager() {
        this.sounds = new ArrayList<>();
        this.audioPlayerManager = new DefaultAudioPlayerManager();
        AudioSourceManagers.registerLocalSource(audioPlayerManager);
        AudioSourceManagers.registerRemoteSources(audioPlayerManager);

        this.audioPlayer = audioPlayerManager.createPlayer();
        this.scheduler = new SoundScheduler(this.audioPlayer);
        this.audioPlayer.addListener(scheduler);
    }

    public static SoundManager getInstance() {
        if (instance == null) {
            instance = new SoundManager();
        }

        return instance;
    }

    public void addToCache(String query, AudioTrack track) {
        System.out.println("Added " + query + " to the cache!");
        songCache.put(query, track);
    }

    public boolean isInCache(String query) {
        return songCache.containsKey(query);
    }

    public AudioTrack getFromCache(String query) {
        return songCache.get(query).makeClone();
    }

    public AudioPlayer getAudioPlayer() {
        return audioPlayer;
    }

    public SoundScheduler getScheduler() {
        return scheduler;
    }

    public List<Sound> getSounds() {
        return sounds;
    }

    public Sound getSound(String name) {
        for (Sound sound : sounds) {
            if (sound.name().equals(name)) {
                return sound;
            }
        }

        return null;
    }

    public AudioPlayerManager getAudioPlayerManager() {
        return audioPlayerManager;
    }

    public AudioPlayerSendHandler getSendHandler() {
        return new AudioPlayerSendHandler(this.audioPlayer);
    }

    public void addSound(Sound sound) {
        sounds.add(sound);
    }

    public boolean hasSound(String name) {
        for (Sound sound : sounds) {
            if (sound.name().equals(name)) {
                return true;
            }
        }

        return false;
    }

    public void loadSounds() {
        File soundsFolder = new File("sounds");

        if (soundsFolder.exists()) {
            File[] sounds = soundsFolder.listFiles();

            if (sounds != null) {
                for (File soundFolder : sounds) {
                    if (soundFolder.isDirectory()) {
                        String name = soundFolder.getName();
                        File soundFile = new File(soundFolder, name + ".mp3");

                        if (soundFile.exists()) {
                            Sound sound = new Sound(name, soundFile);

                            addSound(sound);
                        }
                    }
                }
            }
        }
    }

    public boolean shouldShuffle() {
        return shuffle;
    }

    public void setShuffle(boolean shuffle) {
        this.shuffle = shuffle;
    }

    public boolean shouldTTS() {
        return ttsAnnouncement;
    }

    public void setTTS(boolean ttsAnnouncement) {
        this.ttsAnnouncement = ttsAnnouncement;
    }

    public void clearSounds() {
        sounds.clear();
    }

    public boolean shouldLoop() {
        return this.loop;
    }

    public void setLoop(boolean shouldLoop) {
        this.loop = shouldLoop;
    }
}

File: src/main/java/nl/dannyjelsma/soundboardbot/audio/SoundManager.kt
package nl.dannyjelsma.soundboardbot.audio

import com.sedmelluq.discord.lavaplayer.player.AudioPlayer
import com.sedmelluq.discord.lavaplayer.player.AudioPlayerManager
import com.sedmelluq.discord.lavaplayer.player.DefaultAudioPlayerManager
import com.sedmelluq.discord.lavaplayer.source.AudioSourceManagers
import com.sedmelluq.discord.lavaplayer.track.AudioTrack
import java.io.File

object SoundManager {
    val audioPlayer: AudioPlayer
    private val sounds: MutableList<Sound>
    val audioPlayerManager: AudioPlayerManager
    val scheduler: SoundScheduler
    private var shuffle = false
    private var ttsAnnouncement = true
    private val songCache = HashMap<String, AudioTrack>()
    private var loop = false

    init {
        sounds = ArrayList()
        audioPlayerManager = DefaultAudioPlayerManager()
        AudioSourceManagers.registerLocalSource(audioPlayerManager)
        AudioSourceManagers.registerRemoteSources(audioPlayerManager)
        audioPlayer = audioPlayerManager.createPlayer()
        scheduler = SoundScheduler(audioPlayer)
        audioPlayer.addListener(scheduler)
    }

    fun addToCache(query: String, track: AudioTrack) {
        println("Added $query to the cache!")
        songCache[query] = track
    }

    fun getFromCache(query: String): AudioTrack? {
        return songCache[query]?.makeClone()
    }

    fun getSounds(): List<Sound> {
        return sounds
    }

    fun getSound(name: String): Sound? {
        for (sound in sounds) {
            if (sound.name == name) {
                return sound
            }
        }
        return null
    }

    private fun addSound(sound: Sound) {
        sounds.add(sound)
    }

    fun hasSound(name: String): Boolean {
        for (sound in sounds) {
            if (sound.name == name) {
                return true
            }
        }
        return false
    }

    fun loadSounds() {
        val soundsFolder = File("sounds")
        if (soundsFolder.exists()) {
            val sounds = soundsFolder.listFiles()
            if (sounds != null) {
                for (soundFolder in sounds) {
                    if (soundFolder.isDirectory) {
                        val name = soundFolder.name
                        val soundFile = File(soundFolder, "$name.mp3")
                        if (soundFile.exists()) {
                            val sound = Sound(name, soundFile)
                            addSound(sound)
                        }
                    }
                }
            }
        }
    }

    fun shouldShuffle(): Boolean {
        return shuffle
    }

    fun setShuffle(shuffle: Boolean) {
        this.shuffle = shuffle
    }

    fun shouldTTS(): Boolean {
        return ttsAnnouncement
    }

    fun setTTS(ttsAnnouncement: Boolean) {
        this.ttsAnnouncement = ttsAnnouncement
    }

    fun clearSounds() {
        sounds.clear()
    }

    fun shouldLoop(): Boolean {
        return loop
    }

    fun setLoop(shouldLoop: Boolean) {
        loop = shouldLoop
    }
}